File: passman-core/src/buttercup/parse.rs
use super::types::*;
use crate::vault::HistoryItem;
use chrono::{DateTime, Utc};
use std::collections::{HashMap, HashSet};

pub(super) fn is_trash_group(group: &RawGroup) -> bool {
    group
        .a
        .get("bc_group_role")
        .map(|v| v.value.eq_ignore_ascii_case("trash"))
        .unwrap_or(false)
}

fn datetime_from_millis(ts: u64) -> DateTime<Utc> {
    DateTime::from_timestamp_millis(ts as i64)
        .unwrap_or_else(|| DateTime::from_timestamp(0, 0).unwrap())
}

fn get_field_type(attributes: &HashMap<String, RawValue>, property: &str) -> String {
    let key = format!("BC_ENTRY_FIELD_TYPE:{}", property);
    attributes
        .get(&key)
        .map(|v| v.value.clone())
        .filter(|v| !v.is_empty())
        .unwrap_or_else(|| "text".to_string())
}

fn get_property(properties: &HashMap<String, RawValue>, name: &str) -> String {
    properties
        .get(name)
        .map(|v| v.value.clone())
        .unwrap_or_default()
}

pub(super) fn identify_trash_groups(raw: &RawVault) -> (Option<String>, HashSet<String>) {
    let trash_group_id = raw
        .g
        .iter()
        .find(|g| is_trash_group(g))
        .map(|g| g.id.clone());

    let mut trash_group_ids: HashSet<String> = HashSet::new();
    if let Some(tid) = &trash_group_id {
        trash_group_ids.insert(tid.clone());
        collect_descendants(&raw.g, tid, &mut trash_group_ids);
    }

    (trash_group_id, trash_group_ids)
}

fn collect_descendants(groups: &[RawGroup], parent_id: &str, ids: &mut HashSet<String>) {
    for group in groups {
        if group.g == parent_id && !ids.contains(&group.id) {
            ids.insert(group.id.clone());
            collect_descendants(groups, &group.id, ids);
        }
    }
}

pub(super) fn build_groups(
    raw_groups: Vec<RawGroup>,
    trash_group_id: &Option<String>,
    trash_group_ids: &HashSet<String>,
) -> (Vec<ButtercupGroup>, Vec<ButtercupGroup>) {
    let mut groups = Vec::new();
    let mut trash_groups = Vec::new();
    let mut seen_group_ids: HashSet<String> = HashSet::new();

    for group in raw_groups {
        if group.t.is_empty() || !seen_group_ids.insert(group.id.clone()) {
            continue;
        }

        let parent_id = if group.g.is_empty() {
            None
        } else {
            Some(group.g)
        };

        let buttercup_group = ButtercupGroup {
            id: group.id,
            name: group.t,
            parent_id: parent_id.clone(),
        };

        if trash_group_id.as_ref() == Some(&buttercup_group.id) {
            continue;
        }

        if trash_group_ids.contains(&buttercup_group.id) {
            let mut trash_group = buttercup_group;
            if parent_id.as_ref() == trash_group_id.as_ref() {
                trash_group.parent_id = None;
            }
            trash_groups.push(trash_group);
        } else {
            groups.push(buttercup_group);
        }
    }

    (groups, trash_groups)
}

const STANDARD_PROPERTIES: &[&str] = &["title", "username", "password", "URL", "notes"];

fn extract_custom_fields(entry: &RawEntry) -> Vec<ButtercupCustomField> {
    let standard: HashSet<&str> = STANDARD_PROPERTIES.iter().copied().collect();
    let mut fields = Vec::new();
    for (property, raw_value) in &entry.p {
        if standard.contains(property.as_str()) || raw_value.value.is_empty() {
            continue;
        }
        if raw_value.deleted.is_some() {
            continue;
        }
        let field_type = get_field_type(&entry.a, property);
        fields.push(ButtercupCustomField {
            id: format!("{}-cf-{}", entry.id, fields.len()),
            label: property.clone(),
            field_type,
            value: raw_value.value.clone(),
        });
    }
    fields
}

fn extract_history(entry: &RawEntry) -> Vec<HistoryItem> {
    let mut history = Vec::new();
    for (property, raw_value) in &entry.p {
        for hist in &raw_value.history {
            history.push(HistoryItem {
                property: property.clone(),
                value: hist.value.clone(),
                updated_at: datetime_from_millis(hist.updated),
            });
        }
    }
    history
}

pub(super) fn build_entries(
    raw_entries: Vec<RawEntry>,
    trash_group_ids: &HashSet<String>,
    trash_group_id: &Option<String>,
) -> (Vec<ButtercupEntry>, Vec<ButtercupEntry>) {
    let mut entries = Vec::new();
    let mut trash_entries = Vec::new();

    for entry in raw_entries {
        let group_id = if entry.g.is_empty() {
            None
        } else {
            Some(entry.g.clone())
        };
        let id = entry.id.clone();

        let mut fields = extract_custom_fields(&entry);

        let url = get_property(&entry.p, "URL");
        if !url.is_empty() {
            fields.push(ButtercupCustomField {
                id: format!("{}-cf-url", entry.id),
                label: "URL".to_string(),
                field_type: "text".to_string(),
                value: url,
            });
        }
        let notes = get_property(&entry.p, "notes");
        if !notes.is_empty() {
            fields.push(ButtercupCustomField {
                id: format!("{}-cf-notes", entry.id),
                label: "Notes".to_string(),
                field_type: "note".to_string(),
                value: notes,
            });
        }

        let buttercup_entry = ButtercupEntry {
            id,
            group_id: group_id.clone(),
            title: get_property(&entry.p, "title"),
            username: get_property(&entry.p, "username"),
            password: get_property(&entry.p, "password"),
            fields,
            deleted_at: entry.deleted.map(datetime_from_millis),
            history: extract_history(&entry),
        };

        let is_in_trash_group = group_id
            .as_ref()
            .is_some_and(|gid| trash_group_ids.contains(gid));
        let is_trash = entry.deleted.is_some() || is_in_trash_group;

        if is_trash {
            let mut trash_entry = buttercup_entry;
            if group_id.as_ref() == trash_group_id.as_ref() {
                trash_entry.group_id = None;
            }
            trash_entries.push(trash_entry);
        } else {
            entries.push(buttercup_entry);
        }
    }

    (entries, trash_entries)
}

File: pkg/buttercup/parse.go
package buttercup

import (
	"strconv"
	"strings"
	"time"
)

func isTrashGroup(group *RawGroup) bool {
	if v, ok := group.A["bc_group_role"]; ok {
		return strings.EqualFold(v.Value, "trash")
	}
	return false
}

func datetimeFromMillis(ts uint64) time.Time {
	return time.UnixMilli(int64(ts)).UTC()
}

func getFieldType(attributes map[string]RawValue, property string) string {
	key := "BC_ENTRY_FIELD_TYPE:" + property
	if v, ok := attributes[key]; ok && v.Value != "" {
		return v.Value
	}
	return "text"
}

func getProperty(properties map[string]RawValue, name string) string {
	if v, ok := properties[name]; ok {
		return v.Value
	}
	return ""
}

func IdentifyTrashGroups(raw *RawVault) (*string, map[string]bool) {
	var trashGroupID *string
	for _, g := range raw.G {
		if isTrashGroup(&g) {
			id := g.ID
			trashGroupID = &id
			break
		}
	}

	trashGroupIDs := map[string]bool{}
	if trashGroupID != nil {
		trashGroupIDs[*trashGroupID] = true
		collectDescendants(raw.G, *trashGroupID, trashGroupIDs)
	}
	return trashGroupID, trashGroupIDs
}

func collectDescendants(groups []RawGroup, parentID string, ids map[string]bool) {
	for _, g := range groups {
		if g.G == parentID && !ids[g.ID] {
			ids[g.ID] = true
			collectDescendants(groups, g.ID, ids)
		}
	}
}

var standardProperties = map[string]bool{
	"title":    true,
	"username": true,
	"password": true,
	"URL":      true,
	"notes":    true,
}

func extractCustomFields(entry *RawEntry) []ButtercupCustomField {
	var fields []ButtercupCustomField
	for prop, rv := range entry.P {
		if standardProperties[prop] || rv.Value == "" {
			continue
		}
		if rv.Deleted != nil {
			continue
		}
		fieldType := getFieldType(entry.A, prop)
		fields = append(fields, ButtercupCustomField{
			ID:        entry.ID + "-cf-" + strconv.Itoa(len(fields)),
			Label:     prop,
			FieldType: fieldType,
			Value:     rv.Value,
		})
	}
	return fields
}

func extractHistory(entry *RawEntry) []HistoryItem {
	var history []HistoryItem
	for prop, rv := range entry.P {
		for _, hist := range rv.History {
			history = append(history, HistoryItem{
				Property:  prop,
				Value:     hist.Value,
				UpdatedAt: datetimeFromMillis(hist.Updated),
			})
		}
	}
	return history
}

func BuildGroups(rawGroups []RawGroup, trashGroupID *string, trashGroupIDs map[string]bool) ([]ButtercupGroup, []ButtercupGroup) {
	var groups, trashGroups []ButtercupGroup
	seenGroupIDs := map[string]bool{}

	for _, group := range rawGroups {
		if group.T == "" || seenGroupIDs[group.ID] {
			continue
		}
		seenGroupIDs[group.ID] = true

		var parentID *string
		if group.G != "" {
			parentID = &group.G
		}

		bcGroup := ButtercupGroup{
			ID:       group.ID,
			Name:     group.T,
			ParentID: parentID,
		}

		if trashGroupID != nil && *trashGroupID == bcGroup.ID {
			continue
		}

		if trashGroupIDs[bcGroup.ID] {
			trashGroup := bcGroup
			if trashGroupID != nil && trashGroup.ParentID != nil && *trashGroup.ParentID == *trashGroupID {
				trashGroup.ParentID = nil
			}
			trashGroups = append(trashGroups, trashGroup)
		} else {
			groups = append(groups, bcGroup)
		}
	}
	return groups, trashGroups
}

func BuildEntries(rawEntries []RawEntry, trashGroupIDs map[string]bool, trashGroupID *string) ([]ButtercupEntry, []ButtercupEntry) {
	var entries, trashEntries []ButtercupEntry

	for _, entry := range rawEntries {
		var groupID *string
		if entry.G != "" {
			gid := entry.G
			groupID = &gid
		}
		id := entry.ID

		fields := extractCustomFields(&entry)

		url := getProperty(entry.P, "URL")
		if url != "" {
			fields = append(fields, ButtercupCustomField{
				ID:        id + "-cf-url",
				Label:     "URL",
				FieldType: "text",
				Value:     url,
			})
		}
		notes := getProperty(entry.P, "notes")
		if notes != "" {
			fields = append(fields, ButtercupCustomField{
				ID:        id + "-cf-notes",
				Label:     "Notes",
				FieldType: "note",
				Value:     notes,
			})
		}

		var deletedAt *time.Time
		if entry.Deleted != nil {
			t := datetimeFromMillis(*entry.Deleted)
			deletedAt = &t
		}

		bcEntry := ButtercupEntry{
			ID:        id,
			GroupID:   groupID,
			Title:     getProperty(entry.P, "title"),
			Username:  getProperty(entry.P, "username"),
			Password:  getProperty(entry.P, "password"),
			Fields:    fields,
			DeletedAt: deletedAt,
			History:   extractHistory(&entry),
		}

		isInTrashGroup := groupID != nil && trashGroupIDs[*groupID]
		isTrash := entry.Deleted != nil || isInTrashGroup

		if isTrash {
			trashEntry := bcEntry
			if trashGroupID != nil && trashEntry.GroupID != nil && *trashEntry.GroupID == *trashGroupID {
				trashEntry.GroupID = nil
			}
			trashEntries = append(trashEntries, trashEntry)
		} else {
			entries = append(entries, bcEntry)
		}
	}
	return entries, trashEntries
}
